File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        d={'2':"abc",'3':"def",'4':"ghi",'5':"jkl",'6':"mno",'7':"pqrs",'8':"tuv",'9':"wxyz"}
        l=[]
        first=digits[0]
        for i in range(len(d[first])):
            l.append(d[first][i])
        for i in range(1,len(digits)):
            j=0
            length=len(l)
            while(j<length):
                start=l.pop(0)
                for k in range(0,len(d[digits[i]])):
                    l.append(start+d[digits[i]][k])
                j+=1
        return l;
                

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> ans(String digits,int n,List<String> res,String[] combn){
        if(n<0) return res;

        int len=res.size();
        int curr=digits.charAt(n)-'0'-2;
        for(int i=0;i<len;i++){
            for(int j=0;j<combn[curr].length();j++){
                String s=Character.toString(combn[curr].charAt(j))+res.get(i);
                res.add(s);
            }
        }
        
        for(int i=0;i<len;i++){
            res.remove(0);
        }
        return ans(digits,n-1,res,combn);
    }
    public List<String> letterCombinations(String digits) {
        int n=digits.length();
        String[] combn={"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        List<String> res=new ArrayList<>();
        int curr=digits.charAt(n-1)-'0'-2;
        for(int i=0;i<combn[curr].length();i++){
            res.add(Character.toString(combn[curr].charAt(i)));
        }
        res=ans(digits,n-2,res,combn);
        return res;
    }
}